### PR TITLE
Add wrappers for `MPI_Isend` and `MPI_Irecv` with basic handling of `MPI_Request`s

### PR DESCRIPTION
**Pull request by xann16**
Currently `isend`/`irecv` is added along with `MPI_Request` handling via `wait`/`waitall`/`waitany`/`test`/`testall`/`testany` in a procedural way that prohibits, e.g. handling non-contiguous arrays in a way that it is done in current implementation of blocking `recv`. This may be remedied by trying to implement `Request` object in the style of `mpi4py` later, but for now I think it is a good first step to introduce some async operations to the offering.

This also raises general question of how we want our API to look - there are some inconsistancies, e.g. if status returned by `void`-ish methods can be returned, then it is, but when there is a meaningful return value it is only asserted that the call was a success. I tried to stick to no more than one return value, which led to curiosities that may be witnessed in `waitany` and `testany`. If You have better ideas, feel free to voice them. And - in general - it would be good to have some deeper think about basic principles that public wrapper functions would follow, and then appropriately unify them according to these principles to make entire API more consistent.

### CODE REVIEW

**Comment by slayoo** [comment]
Kudos @xann16 !!!
Concerning API design, I'd say for now the two main guidelines were:
- follow the C API as closely as feasible,
- diverge from the C API only if things can be made more Pythonic as a result (but not OOP).
I'll leave some more comments in the code review

File: numba_mpi/api/requests.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+    status = _MPI_Testall(
+        requests.size, requests.ctypes.data, flag.ctypes.data, status_buffer.ctypes.data
+    )
+
+    assert status == 0
+
+    return flag[0] != 0
+
+
+@numba.experimental.jitclass([("value", numba.int32)])
```
Comment (slayoo) [inline numba_mpi/api/requests.py:146]: Numba jit classes have a very poor performance record. Here, it seems we could perhaps use a named tuple with `flag` and `index` fields?

File: numba_mpi/api/requests.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    index = np.empty(1, dtype=np.intc)
+    status = _MPI_Testany(
+        requests.size,
+        requests.ctypes.data,
+        index.ctypes.data,
+        flag.ctypes.data,
+        status_buffer.ctypes.data,
+    )
+
+    assert status == 0
```
Comment (slayoo) [inline numba_mpi/api/requests.py:197]: again, let's leave it to the user to do such asserts

File: numba_mpi/api/requests.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    libmpi,
+)
+
+# helper function to allocate numpy array of request handles
+
+
+@numba.njit
+def create_requests_array(count=1):
+    """Helper function for creating numpy array storing pointers to MPI_Request handles."""
+    return np.empty(count, dtype=_MpiRequestNpType)
```
Comment (slayoo) [inline numba_mpi/api/requests.py:23]: I'd suggest exposing something lik `_MpiRequestNpType` as a part of the public API, and then leave it to the user if the requests pointers would be stored in np.arrays, tuples or other containers

File: tests/api/test_isend_irecv.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+TEST_WAIT_FULL_IN_SECONDS = 0.3
+TEST_WAIT_INCREMENT_IN_SECONDS = 0.1
+
+
+@pytest.mark.parametrize(
+    "isnd, ircv, wait",
+    (
+        (mpi.isend, mpi.irecv, mpi.wait),
+        (mpi.isend.py_func, mpi.irecv.py_func, mpi.wait.py_func),
```
Comment (slayoo) [inline tests/api/test_isend_irecv.py:20]: using `py_func` here ensures that we have `jit` decorators for all tested functions, but we do not ensure that `njit`
to this end, in other test files, we started introducing test-file-defined njitted wrappers like here: https://github.com/numba-mpi/numba-mpi/blob/6fc280addaf4c82f275a89f191bf06ba9a8a8807/tests/api/test_barrier.py#L10

**Review by slayoo — CHANGES_REQUESTED** [review]

**Comment by xann16** [comment]
Thanks for review, @slayoo 

I'll try to provide necessary adjustments soon.

Concerning issues with API, I'll follow Your suggestions and principle to foolow C-style API, although I am not completely on board with that long term. On one hand, I understand it is crucial to leave clients to deal with statuses as they see fit; on the other, however, if consistently applied it will make this API very "ugly", unreadable and non-Pythonic. If we want to be consistent, we should also adjust `size()` and `rank()` to return tuples (I took - admittedly poor - idea of asserting on status from there), but that would be to some detriment to the clarity of code, even in `README` sample or in tests. The same goes for simplicity of use in conditions of `test[all]()` (see tests). I think this would force many clients to write more readable wrappers that allow clearer code, and maybe that is not such a bad thing.

Do not get me wrong, I am not sure what is the best choice here. Both have significant advantages and drawbacks. Just raising discussiont that deserves some attention. For now I'll make changes to fit "as-close-as-c-style" principle, but in such case we need to consider changing `size()` and `rank` to be consistent with other wrappers.

**Comment by david-zwicker** [comment]
I generally agree that it would be nice to follow the C API, but on the other hand it would be nice to use some more pythonic features. Since numba also supports exceptions, I suggest that we only return the expected result from functions and raise exceptions in case of errors. One disadvantage of this method clearly is that we cannot (currently) pass any additional information via exceptions. One alternative I could see is to use a global variable to store the (last) status, but this does not seem very clean, either.

I guess we should open another issue or even project to decide on the handling of return codes.

**Comment by slayoo** [comment]
@xann16 @david-zwicker thanks for feedback!
I'm all for switching to Numba-supported exceptions and phasing out the statuses (and asserts). Just created an issue here: #108 

As @david-zwicker suggests, we can merge this one with statuses; release; and then consistently refactor the whole codebase switching to exceptions and adding relevant tests + an example in README, in a separate PR.

Let's only use asserts within tests or to mark yet-unsupported logic (like non-contiguous array handling).

If the Numba classes (instead of named tuples) are beneficial for clarity of user code, I'm all for it as well - I was just thinking out loud in the review comments.

Thanks again, @xann16 for this mega-important contribution!

File: numba_mpi/api/irecv.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        data.ctypes.data,
+        data.size,
+        _mpi_dtype(data),
+        source,
+        tag,
+        _mpi_addr(_MPI_Comm_World_ptr),
+        request.ctypes.data,
+    )
+
+    assert status == 0
```
[the remark this answers is not in the document]
Comment (slayoo) [reply numba_mpi/api/irecv.py:38]: deferring to another PR that will address exception handling in the whole numba-mpi codebase

File: numba_mpi/api/requests.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    request is completed. Status is currently not handled. Requires
+    'request' parameter to be a c-style pointer to MPI_Request
+    (such as returned by 'isend'/'irecv').
+    """
+
+    status_buffer = create_status_buffer()
+    flag = np.empty(1, dtype=np.intc)
+    status = _MPI_Test(request.ctypes.data, flag.ctypes.data, status_buffer.ctypes.data)
+
+    assert status == 0
```
[the remark this answers is not in the document]
Comment (slayoo) [reply numba_mpi/api/requests.py:138]: deferring to another PR aimed at introducing exceptions

File: numba_mpi/api/requests.py (hunk 2 above)
Comment (slayoo) [reply numba_mpi/api/requests.py:234]: deferring to another PR introducing exceptions

File: numba_mpi/api/irecv.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+_MPI_Irecv.argtypes = send_recv_async_args
+
+
+@numba.njit()
+def irecv(data, source=ANY_SOURCE, tag=ANY_TAG):
+    """Wrapper for MPI_Irecv (only handles contiguous arrays, at least for now).
+    If successful (i.e. result is MPI_SUCCESS), returns c-style pointer to valid
+    MPI_Request handle that may be used with appropriate wait and test functions."""
+
+    assert data.flags.c_contiguous
```
[the remark this answers is not in the document]
Comment (xann16) [reply numba_mpi/api/irecv.py]: done

File: numba_mpi/api/isend.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        data.ctypes.data,
+        data.size,
+        _mpi_dtype(data),
+        dest,
+        tag,
+        _mpi_addr(_MPI_Comm_World_ptr),
+        request.ctypes.data,
+    )
+
+    assert status == 0
```
[the remark this answers is not in the document]
Comment (xann16) [reply numba_mpi/api/isend.py]: - changed result to tuple
- deferring to another PR that will address exception handling in the whole numba-mpi codebase

File: numba_mpi/api/requests.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+
+from numba_mpi.common import (
+    _MpiRequestNpType,
+    _MpiRequestPtr,
+    _MpiStatusPtr,
+    create_status_buffer,
+    libmpi,
+)
+
+# helper function to allocate numpy array of request handles
```
[the remark this answers is not in the document]
Comment (xann16) [reply numba_mpi/api/requests.py]: done

File: numba_mpi/__init__.py [hunk 10]
```diff
@@ -5,9 +5,20 @@
 from .api.barrier import barrier
 from .api.bcast import bcast
 from .api.initialized import initialized
+from .api.irecv import irecv
+from .api.isend import isend
 from .api.operator import Operator
 from .api.rank import rank
 from .api.recv import recv
+from .api.requests import (
+    create_requests_array,
```
[the remark this answers is not in the document]
Comment (xann16) [reply numba_mpi/__init__.py]: As suggested renamed and removed from public API. Added `RequestType` to public API for users to create appropriate numpy arrays to store requests, but now also tuples of requests are supported. See tests for examples of usage.

File: numba_mpi/api/requests.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+
+@numba.njit
+def create_requests_array(count=1):
+    """Helper function for creating numpy array storing pointers to MPI_Request handles."""
+    return np.empty(count, dtype=_MpiRequestNpType)
+
+
+# Wait* functions
```
[the remark this answers is not in the document]
Comment (xann16) [reply numba_mpi/api/requests.py]: done

File: numba_mpi/api/requests.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        index.ctypes.data,
+        status_buffer.ctypes.data,
+    )
+
+    if status > 0:
+        return -status
+    return index[0]
+
+
+# - Test* functions
```
[the remark this answers is not in the document]
Comment (xann16) [reply numba_mpi/api/requests.py]: done

File: numba_mpi/api/requests.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+    'isend'/'irecv').
+    """
+
+    status_buffer = create_status_buffer(requests.size)
+    flag = np.empty(1, dtype=np.intc)
+    status = _MPI_Testall(
+        requests.size, requests.ctypes.data, flag.ctypes.data, status_buffer.ctypes.data
+    )
+
+    assert status == 0
```
[the remark this answers is not in the document]
Comment (xann16) [reply numba_mpi/api/requests.py]: - changed result to tuple
- deferring to another PR that will address exception handling in the whole numba-mpi codebase

File: numba_mpi/api/requests.py (hunk 1 above)
Comment (xann16) [reply numba_mpi/api/requests.py]: - removed jit class, changed result to tuple
- deferring to another PR that will address exception handling in the whole numba-mpi codebase

File: tests/api/test_isend_irecv.py (hunk 4 above)
Comment (xann16) [reply tests/api/test_isend_irecv.py]: done for all relevant instances - especially required for some functions that started to use overloads

File: numba_mpi/api/requests.py (hunk 3 above)
Comment (xann16) [reply numba_mpi/api/requests.py]: done

**Comment by xann16** [comment]
Thanks @slayoo , glad to be of help. I hope to contribute more regularly now, as I promised some time ago. I just need to familiarize a bit with (nicely) reworked architecture. CI pipeline witnessed prolonged horror with my commits under that PR, but I had a chance to reacquaint myself with the package and its specifics once again, and further developments shall not be as painful, I hope.

@david-zwicker - the idea with exceptions is a great middle-ground here between pythonicity and being faithful to procedural, c-style API. I thought about it, but was not sure how numba works with exceptions, so I was hesitant to propose it. Great that You mentioned it.

**Review by slayoo — APPROVED** [review]

**Comment by slayoo** [comment]
Amazing work @xann16 !!!

**Pull request merged**